File: Gaming_Platform/Models/transactions.py
from Database.db_connection import DatabaseConnection

class Transactions:
        @staticmethod
        def create_transaction(id_transactions, user_id, game_id, amount):
            connection = DatabaseConnection.get_connection()
            cursor = connection.cursor()
            try:
                cursor.execute(
                    "INSERT INTO transactions (id_transactions, user_id, game_id, amount, date) VALUES (%s,%s,%s,%s,CURRENT_TIMESTAMP)",
                    (id_transactions, user_id, game_id, amount)
                )
                connection.commit()
                print(f"✅ The Transactions  {id_transactions} was successfully added !")
            except Exception as e:
                print(f"❌ Error creating The Transactions: {e}")
                connection.rollback()
            finally:
                cursor.close()
                DatabaseConnection.return_connection(connection)

        @staticmethod
        def get_all_transactions():
            connection = DatabaseConnection.get_connection()
            cursor = connection.cursor()
            try:
                cursor.execute("SELECT * FROM transactions")
                transactions = cursor.fetchall()
                return transactions
            except Exception as e:
                print(f"❌ Error reading the Transactions {e}")
                return []
            finally:
                cursor.close()
                DatabaseConnection.return_connection(connection)

        @staticmethod
        def update_transaction_balance(id_transactions, amount):
            connection = DatabaseConnection.get_connection()
            cursor = connection.cursor()
            try:
                cursor.execute(
                    "UPDATE transactions SET amount=amount + %s WHERE id_transactions=%s",
                    (amount, id_transactions)
                )
                connection.commit()
            except Exception as e:
                print(f"❌ Error updating amount: {e}")
                connection.rollback()
            finally:
                cursor.close()
                DatabaseConnection.return_connection(connection)

        @staticmethod
        def delete_transaction(id_transactions):
            connection = DatabaseConnection.get_connection()
            cursor = connection.cursor()
            try:
                cursor.execute(
                    "DELETE FROM transactions WHERE id_transactions=%s",
                    (id_transactions,)
                )
                connection.commit()
            except Exception as e:
                print(f"❌ Error deleting Transaction: {e}")
                connection.rollback()
            finally:
                cursor.close()
                DatabaseConnection.return_connection(connection)

File: Gaming_platform/src/Models/Transactions.java
package Models;

import DatabaseConnection.DataConnector;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

public class Transactions {

    public static void createTransaction(int idTransactions, int userId, int gameId, double amount, String dbType) {
        String query = "INSERT INTO transactions (id_transactions, user_id, game_id, amount, date) VALUES (?, ?, ?, ?, CURRENT_TIMESTAMP)";
        try (Connection connection = DataConnector.connect(dbType);
             PreparedStatement stmt = connection.prepareStatement(query)) {

            stmt.setInt(1, idTransactions);
            stmt.setInt(2, userId);
            stmt.setInt(3, gameId);
            stmt.setDouble(4, amount);

            stmt.executeUpdate();
            System.out.println("✅ The transaction " + idTransactions + " was successfully added!");

        } catch (SQLException e) {
            System.out.println("❌ Error creating the transaction: " + e.getMessage());
        }
    }

    public static void getAllTransactions(String dbType) {
        String query = "SELECT * FROM transactions";
        try (Connection connection = DataConnector.connect(dbType);
             PreparedStatement stmt = connection.prepareStatement(query);
             ResultSet rs = stmt.executeQuery()) {

            while (rs.next()) {
                int id = rs.getInt("id_transactions");
                int userId = rs.getInt("user_id");
                int gameId = rs.getInt("game_id");
                double amount = rs.getDouble("amount");
                String date = rs.getString("date");

                System.out.println("💳 Transaction ID: " + id + " | User ID: " + userId + " | Game ID: " + gameId + " | Amount: $" + amount + " | Date: " + date);
            }

        } catch (SQLException e) {
            System.out.println("❌ Error reading transactions: " + e.getMessage());
        }
    }

    public static void updateTransactionAmount(int idTransactions, double amount, String dbType) {
        String query = "UPDATE transactions SET amount = amount + ? WHERE id_transactions = ?";
        try (Connection connection = DataConnector.connect(dbType);
             PreparedStatement stmt = connection.prepareStatement(query)) {

            stmt.setDouble(1, amount);
            stmt.setInt(2, idTransactions);

            int rowsAffected = stmt.executeUpdate();
            if (rowsAffected > 0) {
                System.out.println("✅ Transaction " + idTransactions + " amount updated!");
            } else {
                System.out.println("⚠️ Transaction with ID " + idTransactions + " does not exist.");
            }

        } catch (SQLException e) {
            System.out.println("❌ Error updating amount: " + e.getMessage());
        }
    }

    public static void deleteTransaction(int idTransactions, String dbType) {
        String query = "DELETE FROM transactions WHERE id_transactions = ?";
        try (Connection connection = DataConnector.connect(dbType);
             PreparedStatement stmt = connection.prepareStatement(query)) {

            stmt.setInt(1, idTransactions);

            int rowsAffected = stmt.executeUpdate();
            if (rowsAffected > 0) {
                System.out.println("✅ Transaction " + idTransactions + " deleted!");
            } else {
                System.out.println("⚠️ Transaction with ID " + idTransactions + " does not exist.");
            }

        } catch (SQLException e) {
            System.out.println("❌ Error deleting transaction: " + e.getMessage());
        }
    }
}
